vslib/src/sai_vs_switch.cpp

```diff
@@ -1,5 +1,287 @@
 #include "sai_vs.h"
 #include "sai_vs_internal.h"
+#include "sai_vs_state.h"
+
+#include "swss/selectableevent.h"
+#include "swss/netmsg.h"
+#include "swss/netlink.h"
+#include "swss/select.h"
+#include "swss/netdispatcher.h"
+
+#include <netlink/route/link.h>
+#include <netlink/route/addr.h>
+#include <linux/if.h>
+
+#include <thread>
+
+/**
+ * @brief Get SwitchState by switch id.
+ *
+ * Function will get shared object for switch state.  This function is thread
+ * safe and it's only intended to use inside threads.
+ *
+ * @param switch_id Switch ID
+ *
+ * @return SwitchState object or null ptr if not found.
+ */
+std::shared_ptr<SwitchState> vs_get_switch_state(
+        _In_ sai_object_id_t switch_id)
+{
+    MUTEX();
+
+    SWSS_LOG_ENTER();
+
+    auto it = g_switch_state_map.find(switch_id);
+
+    if (it == g_switch_state_map.end())
+    {
+        return nullptr;
+    }
+
+    return it->second;
+}
+
+void update_port_oper_status(
+        _In_ sai_object_id_t port_id,
+        _In_ sai_port_oper_status_t port_oper_status)
+{
+    MUTEX();
+
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr;
+
+    attr.id = SAI_PORT_ATTR_OPER_STATUS;
+    attr.value.s32 = port_oper_status;
+
+    sai_status_t status = vs_generic_set(SAI_OBJECT_TYPE_PORT, port_id, &attr);
+
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("failed to update port status %s: %s",
+                sai_serialize_object_id(port_id).c_str(),
+                sai_serialize_port_oper_status(port_oper_status).c_str());
+    }
+}
+
+class LinkMsg : public swss::NetMsg
+{
+    public:
+
+        LinkMsg(sai_object_id_t switch_id)
+            : m_switch_id(switch_id)
+        {
+        }
+
+        virtual void onMsg(int nlmsg_type, struct nl_object *obj)
+        {
+            if (nlmsg_type != RTM_NEWLINK)
+            {
+                SWSS_LOG_WARN("unsupported nlmsg_type: %d", nlmsg_type);
+                return;
+            }
+
+            // new link
+
+            struct rtnl_link *link = (struct rtnl_link *)obj;
+
+            int             if_index = rtnl_link_get_ifindex(link);
+            unsigned int    if_flags = rtnl_link_get_flags(link); // IFF_LOWER_UP and IFF_RUNNING
+            const char*     if_name  = rtnl_link_get_name(link);
+
+            if (if_name[0] != 'v')
```

Comment:
>'v' [](start = 30, length = 3)

this 'v' is used in multiple places, can we put this as a macro, so that if we want to change, we can change only one place.